### PR TITLE
Bean Validation support in reactive routes

### PR DESCRIPTION
**Pull request by cescoffier**
Implement the support of bean validation with reactive routes:

* parameters are checked
* synchronous returns objects, and `Uni<X>` are validated - only `@Valid` is supported in this case

Multis are not supported, as the use case is not clear. 

﻿Ths PR also fixes the Quarkus Error Handler. It  was not enforcing the status code that could have been configured in the context.

### CODE REVIEW

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java [hunk 1]
```diff
@@ -4,21 +4,10 @@
...
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.LinkedHashSet;
-import java.util.LinkedList;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (geoand) [inline extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java:9]: This should probably be avoided :)

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/HandlerDescriptor.java [hunk 2]
```diff
@@ -31,6 +35,34 @@ boolean isReturningMulti() {
...
+    /**
+     * @return {@code true} if the method is annotated with a constraint or {@code @Valid} or any parameter has such kind of
+     *         annotation.
+     */
+    boolean requireValidation() {
+        return method.annotations().stream()
+                .anyMatch(new Predicate<AnnotationInstance>() {
+                    @Override
+                    public boolean test(AnnotationInstance ai) {
+                        return ai.name().toString().startsWith("javax.validation");
```
Comment (gsmet) [inline extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/HandlerDescriptor.java:47]: Hmmm, no, that won't work :). We have `org.hibernate` specific constraints and users develop custom constraints.

Maybe I should add a build item that can send you the list of constraints?

File: extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/mutiny/JsonMultiRouteTest.java [hunk 3]
```diff
@@ -27,37 +25,38 @@
...
+        //                .header("content-type", "application/json");
+        //
+        //        when().get("/people-content-type").then().statusCode(200)
+        //                .body("size()", is(3))
+        //                .body("[0].name", is("superman"))
+        //                .body("[1].name", is("batman"))
+        //                .body("[2].name", is("spiderman"))
+        //                .header("content-type", "application/json;charset=utf-8");
+
+        System.out.println(when().get("/failure").asString());
```
Comment (gsmet) [inline extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/mutiny/JsonMultiRouteTest.java:59]: All this makes me a bit nervous?

**Review by gsmet — CHANGES_REQUESTED** [review]
I will work this afternoon on publishing the list of detected constraints.

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/HandlerDescriptor.java (hunk 2 above)
Comment (mkouba) [reply extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/HandlerDescriptor.java:47]: BTW I also like streams but a loop returning `true` once a constraint annotation is found would be more readable ;-)

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java [hunk 4]
```diff
@@ -574,7 +602,32 @@ void implementInvoke(HandlerDescriptor descriptor, BeanInfo bean, MethodInfo met
...
+            if (value != null) {
+                invoke.assign(res, value);
+            }
+        } else {
+            TryBlock block = invoke.tryBlock();
+            ResultHandle value = block.invokeVirtualMethod(methodDescriptor, beanInstanceHandle, paramHandles);
+            if (value != null) {
+                block.assign(res, value);
+            }
+            CatchBlockCreator caught = block.addCatch("javax.validation.ConstraintViolationException");
```
Comment (mkouba) [inline extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java:625]: I'd suggest to use constants for repeating strings...

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java [hunk 5]
```diff
@@ -468,8 +479,16 @@ private String generateHandler(HandlerDescriptor desc, BeanInfo bean, MethodInfo
                     .setModifiers(ACC_PRIVATE | ACC_FINAL);
         }
 
-        implementConstructor(bean, invokerCreator, beanField, contextField, containerField);
-        implementInvoke(desc, bean, method, invokerCreator, beanField, contextField, containerField, transformedAnnotations,
+        FieldCreator validatorField = null;
+        if (desc.isProducedResponseValidated()) {
+            // Is the produced item needs to be validated, we inject the Validator
+            validatorField = invokerCreator.getFieldCreator("validator", "javax.validation.Validator")
+                    .setModifiers(ACC_PUBLIC);
```
Comment (mkouba) [inline extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java:486]: Why public?

**Review by mkouba — CHANGES_REQUESTED** [review]
This PR contains a lot of unncessary formatting changes which makes it harder to review :-(

But overall it looks good and is definitely a useful addition!

File: extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/ValidationSupport.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                    break;
+                }
+            }
+            // If not JSON just fails.
+            rc.fail(status, ex);
+        }
+    }
+
+    public static void log(String s) {
+        System.out.println("[LOG] - " + s);
```
Comment (gsmet) [inline extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/ValidationSupport.java:89]: Hmmm, should we really use `System.out.println` here and below?

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java [hunk 7]
```diff
@@ -468,8 +479,16 @@ private String generateHandler(HandlerDescriptor desc, BeanInfo bean, MethodInfo
                     .setModifiers(ACC_PRIVATE | ACC_FINAL);
         }
 
-        implementConstructor(bean, invokerCreator, beanField, contextField, containerField);
-        implementInvoke(desc, bean, method, invokerCreator, beanField, contextField, containerField, transformedAnnotations,
+        FieldCreator validatorField = null;
+        if (desc.isProducedResponseValidated()) {
+            // Is the produced item needs to be validated, we inject the Validator
```
Comment (mkouba) [inline extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java:484]: "Is the produced item" -> "If the produced item"

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/Methods.java [hunk 8]
```diff
@@ -168,6 +173,17 @@
     static final MethodDescriptor OPTIONAL_OF_NULLABLE = MethodDescriptor
             .ofMethod(Optional.class, "ofNullable", Optional.class, Object.class);
 
+    static final MethodDescriptor GET_VALIDATOR = MethodDescriptor.ofMethod(ValidationSupport.class, "getValidator",
```
Comment (mkouba) [inline extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/Methods.java:176]: It's good to start the name of a constant with the class/domain name, e.g. `VALIDATION_GET_VALIDATOR`, `SET_IS_EMPTY`...

File: extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/ValidationSupport.java (hunk 6 above)
Comment (mkouba) [reply extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/ValidationSupport.java:89]: Are the log methods even used somewhere?

File: extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/validation/SyncValidationTest.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+                .header("Accept", "application/json")
+                .when()
+                .get("/invalid")
+                .then()
+                .statusCode(500)
+                .body("title", containsString("Constraint Violation"))
+                .body("status", is(500))
+                .body("details", containsString("validation constraint violations"))
+                .body("violations[0].field", containsString("name"))
+                .body("violations[0].message", containsString("length"));
```
Comment (mkouba) [inline extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/validation/SyncValidationTest.java:57]: These tests will fail on systems with non-EN locale...

File: extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/mutiny/JsonMultiRouteTest.java (hunk 3 above)
Comment (cescoffier) [reply extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/mutiny/JsonMultiRouteTest.java:59]: forgot to uncomment after debugging

File: extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/ValidationSupport.java (hunk 6 above)
Comment (cescoffier) [reply extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/ValidationSupport.java:89]: unused, was only there for debugging

File: extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/validation/SyncValidationTest.java (hunk 9 above)
Comment (mkouba) [reply extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/validation/SyncValidationTest.java:57]: I think the default locale for a JVM is derived from the OS and in my case, it's `cs_CZ.UTF-8`. So I get czech messages for the constraints violations...

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/HandlerDescriptor.java [hunk 10]
```diff
@@ -31,6 +33,34 @@ boolean isReturningMulti() {
...
     }
 
+    /**
+     * @return {@code true} if the method is annotated with a constraint or {@code @Valid} or any parameter has such kind of
+     *         annotation.
+     */
+    boolean requireValidation() {
+        for (AnnotationInstance annotation : method.annotations()) {
+            String name = annotation.name().toString();
+            if (name.startsWith("javax.validation") || name.startsWith("org.hibernate.validator.constraints")) {
```
Comment (cescoffier) [inline extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/HandlerDescriptor.java:43]: @gsmet @mkouba Not great, but a temporary approach until we get the build item.

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java [hunk 11]
```diff
@@ -468,8 +489,16 @@ private String generateHandler(HandlerDescriptor desc, BeanInfo bean, MethodInfo
                     .setModifiers(ACC_PRIVATE | ACC_FINAL);
         }
 
-        implementConstructor(bean, invokerCreator, beanField, contextField, containerField);
-        implementInvoke(desc, bean, method, invokerCreator, beanField, contextField, containerField, transformedAnnotations,
+        FieldCreator validatorField = null;
+        if (desc.isProducedResponseValidated()) {
+            // If the produced item needs to be validated, we inject the Validator
+            validatorField = invokerCreator.getFieldCreator("validator", Methods.VALIDATION_VALIDATOR)
+                    .setModifiers(ACC_PUBLIC | ACC_FINAL);
```
Comment (cescoffier) [inline extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java:496]: The field is accessed from another class (a callback). Made it public final after discussion with @mkouba

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java (hunk 1 above)
Comment (cescoffier) [reply extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java]: Fixed.

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/HandlerDescriptor.java (hunk 2 above)
Comment (cescoffier) [reply extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/HandlerDescriptor.java]: I've changed the code to use regular loops. 

About the name, I've done a temporary fix until we get the built item.

File: extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/mutiny/JsonMultiRouteTest.java (hunk 3 above)
Comment (cescoffier) [reply extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/mutiny/JsonMultiRouteTest.java]: Fixed.

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java (hunk 4 above)
Comment (cescoffier) [reply extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java]: Fixed

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java (hunk 5 above)
Comment (cescoffier) [reply extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java]: Switch to `public final` as the field is used from a different class (a callback class).

File: extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/ValidationSupport.java (hunk 6 above)
Comment (cescoffier) [reply extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/ValidationSupport.java]: Removed

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/Methods.java (hunk 8 above)
Comment (cescoffier) [reply extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/Methods.java]: Fixed.

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java (hunk 7 above)
Comment (cescoffier) [reply extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java]: Fixed.

File: extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/validation/SyncValidationTest.java (hunk 9 above)
Comment (cescoffier) [reply extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/validation/SyncValidationTest.java]: Checking for nonblank.

**Review by mkouba — APPROVED** [review]
Looks good but we should probably wait for the new build item from @gsmet and then merge ;-).

**Comment by gsmet** [comment]
@cescoffier I merged the BuildItem. Can you rebase and make use of it?

**Comment by cescoffier** [comment]
@gsmet done.

**Comment by mkouba** [comment]
> I merged the BuildItem. Can you rebase and make use of it?

@gsmet Is there a reason why not to use ``DotName``s directly in the build item? `DotName#equals()` is optimized and we could avoid all the `DotName#toString()` invocations which are IMO unnecessary in most cases...

**Comment by cescoffier** [comment]
Anything else required on this?

**Comment by gsmet** [comment]
@mkouba it's mostly to be consistent with other build items. AFAIK most of our build items are using `String` for class names and not `DotName`.

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/HandlerDescriptor.java [hunk 12]
```diff
@@ -31,6 +36,39 @@ boolean isReturningMulti() {
...
+            String name = annotation.name().toString();
+            if (validationAnnotations.getAllAnnotations().contains(name)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * @return {@code true} if the method is annotated with a constraint or {@code @Valid}.
```
Comment (gsmet) [inline extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/HandlerDescriptor.java:57]: This is incorrect, you only test for Valid.

**Review by gsmet — CHANGES_REQUESTED** [review]
@cescoffier just 2 minor comments: the one from Martin for the typo in the doc and mine for an incorrect comment.

Apart from that, it looks good to go and a very nice addition for 1.9!

**Comment by mkouba** [comment]
> it's mostly to be consistent with other build items.

@gsmet I don't think it's very practical, esp. in this case where you first [convert dot names to strings](https://github.com/quarkusio/quarkus/blob/master/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/HibernateValidatorProcessor.java#L208) and then the consumer [converts dot name to string](https://github.com/quarkusio/quarkus/pull/11864/files#diff-2947d0a9862578678781c8f97e31c955R48) again. But it's a very minor "snag" ;-).

> AFAIK most of our build items are using String for class names and not DotName.

Those I've created usually don't ;-).

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/HandlerDescriptor.java (hunk 12 above)
Comment (cescoffier) [reply extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/HandlerDescriptor.java]: Fixed

File: docs/src/main/asciidoc/reactive-routes.adoc [hunk 13]
```diff
@@ -475,6 +475,40 @@ id: 3
...
+
+[source,java]
+----
+@Route(...)
+@Valid Uni<Person> createPerson(@Body @Valid Person person, @NonNull @Param("id") String primaryKey) {
+  // ...
+}
+----
+
+If the item produces by the route does not pass the validation, it returns a HTTP 500 response.
```
[the remark this answers is not in the document]
Comment (cescoffier) [reply docs/src/main/asciidoc/reactive-routes.adoc]: Fixed.

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/HandlerDescriptor.java (hunk 10 above)
Comment (cescoffier) [reply extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/HandlerDescriptor.java]: This has been fixed with the new Build Items provided by Hibernate Validator.

**Comment by cescoffier** [comment]
CI issue not related to this PR (MongodbPanacheMockingTest).

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Merged, thanks!

**Pull request merged**